### ISSUE TITLE
Add GeneratorStrategy::getJavaMemberOverride, getJavaSetterOverride, getJavaGetterOverride, getJavaMethodOverride

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#10776]
When using `<recordImplements/>` and other generator strategy configurations, e.g. to extend interface `X`:

```kotlin
interface X {
    var value: Int?
}
```

Then, there's a compilation error in the implementing records, pojos, etc. because of a missing `override` modifier, e.g. on:

```kotlin
    /* should have override */ var value: Int?
        set(value) = set(1, value)
        get() = get(1) as Int?
```

There needs to be a way to insert this modifier, either automatically (difficult, as we don't know the contents of `<recordImplements/>`), or using additional strategy configuration (possibly not very user friendly)

----
As reported here by @mptardy:

- https://groups.google.com/g/jooq-user/c/EvBzkxYfV-8
- https://github.com/thomastardy/jooq-kotlin-code-generation

**Comment by lukaseder** [comment]
So, this seems to be about `<recordImplements/>` and `<pojoImplements/>`. How can jOOQ's code generator detect the necessity of generating the `override` modifier in those cases? I don't think it's possible, @mptardy (and in Java, it's not necessary)...

**Comment by mptardy** [comment]
Hmmm, the only solution that comes to my mind would be to use reflection to analyze the interfaces and generate the override modifier where appropriate. But this is definitely not an easy solution.
Are you doing something similar during the code generation already?

**Comment by lukaseder** [comment]
> Hmmm, the only solution that comes to my mind would be to use reflection to analyze the interfaces and generate the override modifier where appropriate. But this is definitely not an easy solution.

We probably don't have your interfaces on the classpath. If we did, then it would be easy to introspect them. We do that already in case there's a conflict with some method from a known super class from jOOQ, in case of which we append a `_` to disambiguate it...

> @lukaseder, are you doing something similar during the code generation already?

We generate `override` modifiers when you add `<interfaces>true</interfaces>`. But those interfaces may not be the ones you want... Other than that, I don't think we do something like that already.

**Comment by lukaseder** [comment]
OK, so there are three options:

1. The interface is on the classpath, in case of which we can use reflection. That should be rather straightforward
2. The interface is not on the classpath, which means it's probably in the same project (not necessarily). If it is, we could use kibble to parse the Kotlin sources and hack our way around its contents: https://github.com/evanchooly/kibble
3. If all else fails, then we probably need additional config containing this information...

**Comment by mptardy** [comment]
If you would anyways support the third option, I would go with that. It is not as convenient as the other options but would work in all cases.

**Comment by lukaseder** [comment]
> If you would anyways support the third option, I would go with that. It is not as convenient as the other options but would work in all cases.

Well, it's definitely the easiest (and thus quickest) to implement, e.g. even for 3.14.1

**Comment by mptardy** [comment]
I would definitely prefer that ;-)

**Comment by lukaseder** [comment]
Trying to answer https://stackoverflow.com/q/66202370/521799, I'm trying to write a test right now that is prevented by this. We'll definitely need some tooling that allows for specifying which generated artifacts should get a Java `@Override` or kotlin `override` annotation / modifier.

**Comment by mcspud** [comment]
@lukaseder I was the one that posted  https://stackoverflow.com/q/66202370/521799, thanks for your response.
I'm not sure if this is informative, but I successfully adapted your solution and have posted its implementation in the SO thread.

Let me know if you need anything from me.

Cheers!

**Comment by lukaseder** [comment]
Thanks for answering your own question, @mcspud. That will be very useful for others, in the future!

**Comment by arash-kamangir** [comment]
Hi @lukaseder, is there any workaround about this issue? Thanks, Arash.

**Comment by lukaseder** [comment]
I cannot think of one @arash-kamangir, other than patching generated sources with some post processing, of course (e.g. regex search/replace)

**Comment by lukaseder** [comment]
Another case of this:

- https://github.com/jOOQ/jOOQ/issues/16081

Looking again at the options I presented here:

- https://github.com/jOOQ/jOOQ/issues/10776#issuecomment-713674663

I think the first thing to implement is some way of allowing the `GeneratorStrategy` to tell the `KotlinGenerator` if a property needs the `override` modifier (for whatever reason).

Maybe, in the future, we can add a clever implementation that can derive the result from actual code. Until then, configuring something like `id|created|deleted|updated` (a regular expression, as always), might be good enough.

**Comment by reubenfirmin** [comment]
Agreed, I like the configuration option as a pragmatic initial approach. 

Another possibility: you could require the interfaces/superclasses to live along with the configuration project, and then just copy them into the generated code, whereever that will be. That would also allow code-based configurations to refer to the interface types directly rather than as strings.

**Comment by lukaseder** [comment]
> Another possibility: you could require the interfaces/superclasses to live along with the configuration project, and then just copy them into the generated code, whereever that will be. That would also allow code-based configurations to refer to the interface types directly rather than as strings.

People will have all sorts of opinions on what these interfaces should be able to do, including extending other interfaces. I don't really want to go down that rabbit hole.

**Comment by reubenfirmin** [comment]
Hm, OK. What about passing properties through GeneratorStrategy also and allowing users the chance to modify name, overrides, etc etc?

**Comment by lukaseder** [comment]
> What about passing properties through GeneratorStrategy also and allowing users the chance to modify name, overrides, etc etc?

Properties don't scale and are terrible to document. They're always a hack. How would you even encode all these things. Also, why would users need another mechanism to modify names, etc. etc. (what are etc.?) when there's already `GeneratorStrategy`.

**Comment by lukaseder** [comment]
Of course, a workaround is to avoid the problem like this:

```kotlin
interface I {
    fun m_(): Int
}

class C : I {
    fun m(): Int = 1
    // Implementation generated in a "custom code section"
    override fun m_(): Int = m()
}
```

See: https://www.jooq.org/doc/latest/manual/code-generation/codegen-custom-code/

**Comment by reubenfirmin** [comment]
> why would users need another mechanism to modify names

Not that I necessarily need to, but _can_ I distinguish tables and fields/properties using Definition? As far as I can see the naming strategy is fairly broad-brush.

```
    override fun getJavaIdentifier(definition: Definition): String {
        val identifier = super.getJavaIdentifier(definition)
        return identifier.split('_')
            .joinToString("") { it.lowercase().capitalized() }
    }
```

>  when there's already GeneratorStrategy

Well, right, I'm suggesting extending GeneratorStrategy. Suppose you had:

```
    // not necessarily "property". call it whatever
    override fun getFieldMetadata(definition: Definition): FieldMetadata<...> {
          val metadata = super.getFieldMetdata(definition)
          if (definition.name == "Id") { // is this pre or post "getJavaIdentifier"? 
               metadata.requiresOverride = true
          }
          return metadata
    }
```

**Comment by lukaseder** [comment]
> > why would users need another mechanism to modify names
> 
> Not that I necessarily need to, but _can_ I distinguish tables and fields/properties using Definition? As far as I can see the naming strategy is fairly broad-brush.

I'm sorry, I misunderstood. I thought you were referring to system properties. I don't have a clear picture regarding the implementation of this yet, including whether this is a feature of properties (kotlin specific) or setters/getters (code generation internal specific), or what other options there are... So, I cannot yet comment on how to do this.

> Well, right, I'm suggesting extending GeneratorStrategy. Suppose you had: [...]

Looks like a ton of complex features that are needed only for this very specific purpose right now, which is dangerous because most of this `FieldMetadata` will be overengineered, or useless, should other requirements arise in the future. It's hard to get such APIs right.

**Comment by lukaseder** [comment]
> I don't have a clear picture regarding the implementation of this yet

I mean, an implementation in `JavaGenerator` rather than `GeneratorStrategy` is just as likely

**Comment by lukaseder** [comment]
A hack is to override `JavaGenerator.printKotlinSetterAnnotation()` and add the `override` modifier to the end of the default implementation.

**Comment by reubenfirmin** [comment]
Fair enough, I always appreciate yagni as a reason.

I'll take a look at codegen in the meantime. I was just trying to find the right place in that class - thanks for the pointer.

**Comment by lukaseder** [comment]
I think I'll go with 3 new methods on `GeneratorStrategy`:

- `getJavaMemberOverride(Definition, Mode): boolean`
- `getJavaSetterOverride(Definition, Mode): boolean`
- `getJavaGetterOverride(Definition, Mode): boolean`

This isn't kotlin specific. In Java and Scala, users may also like to have jOOQ generate the `override` modifier, even if it's more cosmetic there.

**Comment by lukaseder** [comment]
The corresponding matcher strategies will be added separately:

- https://github.com/jOOQ/jOOQ/issues/16093

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.20.0

**Comment by lukaseder** [comment]
I like how this turned out, especially using the new matchers. From the integration tests:

```xml
<strategy>
    <matchers>
        <tables>
            <table>
                <expression>OVERRIDE_2</expression>
                <recordImplements>org.jooq.codegen.test.IC</recordImplements>
                <pojoImplements>org.jooq.codegen.test.IC</pojoImplements>
            </table>
        </tables>
        <fields>
            <field>
                <expression>.*\.OVERRIDE_2\.(ID|DC|DM)</expression>
                <recordMemberOverride>true</recordMemberOverride>
                <pojoMemberOverride>true</pojoMemberOverride>
            </field>
        </fields>
    </matchers>
</strategy>
```
